Fixed #30864 -- Doc'd classproperty decorator.

ProProgrammer opened the pull request:
- Added documentation for classproperty and docstring to class definition
- Add contributor name to AUTHORS

Ticket: [#30864](https://code.djangoproject.com/ticket/30864)

_PS: Signed ICLA sent via email to cla@djangoproject.com_

ProProgrammer commented:
**Rendered documentation for `classproperty`**

![Screenshot 2020-03-28 at 8 13 54 PM](https://user-images.githubusercontent.com/3947424/77825676-d12d5780-7130-11ea-891a-2c3139a4ddf4.png)


**Rendered docstring in class definition for `classproperty`**

![Screenshot 2020-03-28 at 8 16 13 PM](https://user-images.githubusercontent.com/3947424/77825700-13569900-7131-11ea-83f3-e51eb74f7768.png)

felixxm commented on django/utils/functional.py:
```diff
@@ -50,6 +50,31 @@ def __get__(self, instance, cls=None):
 class classproperty:
+    """
+    Like @property but allows the property to be accessed directly from the class
+    without requiring an instance of the class, example usage::
```
I don't see much value in adding an example in this docstring. Please remove it.

felixxm commented on django/utils/functional.py:
```diff
@@ -50,6 +50,31 @@ def __get__(self, instance, cls=None):
 
 class classproperty:
+    """
+    Like @property but allows the property to be accessed directly from the class
```
Maybe:
```
"""A property that can be accessed directly from the class."""
````

felixxm commented on docs/ref/utils.txt:
```diff
@@ -556,6 +556,36 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
+        SomeModel.user_method     # returns function reference
+        # <function <app_name>.models.SomeModel.user_method(self)>
+
+    .. versionchanged:: 3.1
```
This annotation is not necessary, please remove it.

felixxm commented on docs/ref/utils.txt:
```diff
@@ -556,6 +556,36 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
         z = person.friends         # does not call
         x is z                     # is True
 
+.. class:: classproperty
```
`classproperty` -> `classproperty(method=None)`

ProProgrammer replied:
Thank you @felixxm, curious why do we mention `(method=None)`, is this because this class doesn't define any method for user to define/overwrite or is there some other thought process behind this?

felixxm replied:
Because that's the signature of this class, see `classproperty.__init__()`.

ProProgrammer replied:
Thanks, got it. So we mention the signature in docs. This is helpful.

**Quick note,** I wasn't able to see the signature using `classproperty.__init__()`, it worked for me using `inspect.signature()`

```
from django.utils.functional import classproperty
print(classproperty.__init__())

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-871ae726864b> in <module>
      2 obj = classproperty()
      3 
----> 4 print(classproperty.__init__())

TypeError: __init__() missing 1 required positional argument: 'self'
```

```
from django.utils.functional import classproperty
obj = classproperty()

print(obj.__init__())
None
```

Finally with `inspect.signature`
```
import inspect
print(inspect.signature(classproperty))

(method=None)
```

felixxm commented on docs/ref/utils.txt:
```diff
@@ -556,6 +556,36 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
         x is z                     # is True
 
+.. class:: classproperty
+
```
`versionadded` is missing:
```
.. class:: classproperty(method=None)

.. versionadded:: 3.1
```

ProProgrammer replied:
At all other places, `versionchanged` / `versionadded` annotations were indented as if they were "inside" the code block, so in order to maintain consistency, I have kept this indented too "inside" `class:: classproperty(method=None)` block as follows, please let me know if this is not how it is supposed to be done.

```
.. class:: classproperty(method=None)

    .. versionchanged:: 3.1
```

felixxm commented on docs/ref/utils.txt:
```diff
@@ -556,6 +556,36 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
+    access the property.::
+
+        # the model
+        class SomeModel(models.Model):
```
IMO an example is not necessary.

felixxm commented on docs/ref/utils.txt:
```diff
@@ -556,6 +556,36 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
+
+    The ``@classproperty`` decorator allows the property to be accessed
+    directly from the class without requiring an instance of the class to
+    access the property.::
```
Maybe:
```
Returns a property that can be accessed directly from the class, e.g. from class methods.
```

felixxm commented:
We can also adjust release notes:
```diff
diff --git a/docs/releases/3.1.txt b/docs/releases/3.1.txt
index ecd70feb2e..549b936e4e 100644
--- a/docs/releases/3.1.txt
+++ b/docs/releases/3.1.txt
@@ -554,8 +554,8 @@ Miscellaneous
   ``ETag`` header to responses with an empty
   :attr:`~django.http.HttpResponse.content`.
 
-* ``django.utils.decorators.classproperty()`` decorator is moved to
-  ``django.utils.functional.classproperty()``.
+* ``django.utils.decorators.classproperty()`` decorator is made public and
+  moved to :class:`django.utils.functional.classproperty()`.
 
 * :tfilter:`floatformat` template filter now outputs (positive) ``0`` for
   negative numbers which round to zero.
```

ProProgrammer commented:
Thanks so much @felixxm 🌟 
Addressed all review comments. 

I realised I kinda made your life difficult, since you effectively gave me the whole changeset yourself in code review. Will keep these feedbacks in mind for future changes. 🙏

felixxm commented:
> With this explanation, example might actually be helpful.
> 
> It is probably my lack of Python familiarity but if I was reading **method with a single cls argument** for the first time, I would have felt a little confused and starting searching on Google/StackOverflow to see its usage examples or might have jumped into a shell_plus notebook and start playing around with it to get a grasp of what it means.
> 

Previous example was too generic. Do you have a proposition for short example with a real use case?

ProProgrammer commented:
> > With this explanation, example might actually be helpful.
> > It is probably my lack of Python familiarity but if I was reading **method with a single cls argument** for the first time, I would have felt a little confused and starting searching on Google/StackOverflow to see its usage examples or might have jumped into a shell_plus notebook and start playing around with it to get a grasp of what it means.
> 
> Previous example was too generic. Do you have a proposition for short example with a real use case?

Thanks, what do you think of this?
```
from datetime import datetime

from django.utils.functional import classproperty

class TemplateSelector:

    email_template = {
        'morning': 'morning_email.html',
        'other': 'other_email.html'
    }

    @classproperty
    def template_name(cls) -> str:
        if datetime.now().hour < 12:
            return cls.email_template['morning']
        else:
            return cls.email_template['other']

# Template name can be accessed directly from the cls
# Notice: Since template_name is a property, no parenthesis "()" at the end
TemplateSelector.template_name    # 'morning' or 'other' depending on hour of the day
```

felixxm commented:
`classproperty` is not a basic tool, so we don't need to explain lack of parentheses, we also don't need to repeat a documentation (`Template name can be accessed directly from the cls`). It would be great to add sth with `Person` to keep all examples consistent :thinking:

carltongibson commented on docs/ref/utils.txt:
```diff
@@ -556,6 +556,14 @@ https://web.archive.org/web/20110718035220/http://diveintomark.org/archives/2004
+    The ``@classproperty`` decorator converts the result of a method with a
+    single ``cls`` argument into a property that can be accessed directly from
+    the class.
+
```
**Maybe** it would be worth mentioning `@classmethod` here? 
("Maybe...")

```suggestion
    Similar to :py:func:`classmethod`, the ``@classproperty`` decorator
    converts the result of a method with a single ``cls`` argument into a
    property that can be accessed directly from the class.
```

(Didn't test the rst-link.)

felixxm replied:
Thanks, added.

ProProgrammer replied:
Thank you both for promptly bringing this to closure 😄

felixxm commented:
Merged in 4b146e0c83891fc67a422aa22f846bb7654c4d38.
